sycl/doc/extensions/experimental/sycl_ext_oneapi_spec_constant_length_alloca.asciidoc

```diff
@@ -0,0 +1,195 @@
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== The `private_alloca` function
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+template <typename ElementType, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+private_alloca(kernel_handler &h);
+} // namespace sycl::ext::oneapi::experimental
+----
+
+This extension adds a new `private_alloca` function that can be used to allocate
+a private memory region with capacity for
+`h.get_specialization_constant<SpecName>()` elements of type `ElementType`. The
+returned pointer will be aligned for `ElementType`.
+
+The underlying memory region is automatically freed when the caller to
+`private_alloca` is returned.
+
+In case of private memory exhaustion, the underlying backend must report an
+error in the same fashion as if the allocation size were static.
+
+If this function is called from host context or an unsupported backend, an
+`exception` with the `errc::feature_not_supported` error code must be thrown.
+
+`ElementType` must be a cv-unqualified trivial type. The return memory is
+default initialized.
+
+==== Parameters
+
+`h`:: `sycl::kernel_handler` used to obtain the value of `SpecName`
+
+==== Template Parameters
+
+`ElementType`:: Cv-unqualified trivial type serving as `value_type` of the
+  returned `sycl::multi_ptr`.
+`SpecName`:: `sycl::specialization_id` of integral `value_type`. The allocated
+  memory region has capacity for `h.get_specialization_constant<SpecName>`
+  `ElementType` elements. The default value for the specialization constant must
+  be at least one and the specialization constant must not be set to a value
+  less than one. Setting the specialization constant to a value less than 1 or
+  providing a default value less than 1 is undefined behaviour.
+`DecorateAddress`:: Whether the returned `sycl::multi_ptr` is decorated or not.
+
+==== Return Value
+
+`sycl::private_ptr` to a region of `h.get_specialization_constant<SpecName>()`
+elements of type `ElementType` aligned for such type. The underlying memory
+region will be automatically deallocated when the function from which
+`private_alloca` is called returns.
+
+== Example usage
+
+This non-normative section shows some example usages of the extension.
+
+[source,c++]
+----
+constexpr specialization_id<int> size(1);
+
+class Kernel;
+
+SYCL_EXTERNAL void impl(const float *in, float *out, size_t n,
+                        decorated_private_ptr<float> ptr);
+
+void run(queue q, const float *in, float *out, size_t n) {
+  q.submit([&](handler &h) {
+    h.set_specialization_constant<size>(n);
+    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler h) {
+      // Allocate memory for 'n' 'float's
+      auto ptr = private_alloca<float, size, access::decorated::yes>(h);
+      // Use pointer in implementation
+      impl(in, out, h.get_specialization_constant<size>(), ptr);
```

Comment:
Reusing the name `h` is legal, but confusing in this case.  I suggest renaming the kernel handler:

```suggestion
    h.parallel_for<Kernel>(n, [=](id<1> i, kernel_handler kh) {
      // Allocate memory for 'n' 'float's
      auto ptr = private_alloca<float, size, access::decorated::yes>(kh);
      // Use pointer in implementation
      impl(in, out, kh.get_specialization_constant<size>(), ptr);
```